lib/s3_inventory.rb

```diff
@@ -0,0 +1,192 @@
+            connection.put_copy_data("#{row[CSV_KEY_INDEX]},#{row[CSV_ETAG_INDEX]}\n")
+          end
+        end
+
+        missing_uploads = model.joins("LEFT JOIN #{table_name} ON #{table_name}.etag = #{model.table_name}.etag").where("#{table_name}.etag is NULL")
+        missing_count = missing_uploads.count
+
+        if missing_count > 0
+          missing_uploads.find_each do |upload|
+            puts upload.url
+          end
+
+          puts "#{missing_count} of #{model.count} #{model.name.underscore.pluralize} are missing"
+        end
+      ensure
+        connection.exec("DROP TABLE #{table_name}") unless connection.nil?
+      end
+    end
+  end
+
+  def copy_archive_to_tmp_directory
+    log "Downloading archive to tmp directory..."
+    failure_message = "Failed to download archive to tmp directory."
+
+    @s3_helper.download_file(file.key, @archive_filename, failure_message)
+  end
+
+  def unzip_archive
+    log "Unzipping archive, this may take a while..."
+
+    FileUtils.cd(@tmp_directory) do
+      Discourse::Utils.execute_command('gzip', '--decompress', @archive_filename, failure_message: "Failed to unzip archive.")
+    end
+  end
+
+  def update_bucket_policy
+    @s3_helper.s3_client.put_bucket_policy(
+      bucket: bucket_name,
+      policy: {
+        "Version": "2012-10-17",
+        "Statement": [
+          {
+            "Sid": "InventoryAndAnalyticsPolicy",
+            "Effect": "Allow",
+            "Principal": { "Service": "s3.amazonaws.com" },
+            "Action": ["s3:PutObject"],
+            "Resource": ["arn:aws:s3:::#{inventory_path}/*"],
+            "Condition": {
+              "ArnLike": {
+                "aws:SourceArn": "arn:aws:s3:::#{bucket_name}"
+              },
+              "StringEquals": {
+                "s3:x-amz-acl": "bucket-owner-full-control"
+              }
+            }
+          }
+        ]
+      }.to_json
+    )
+  end
+
+  def update_bucket_inventory_configuration
+    @s3_helper.s3_client.put_bucket_inventory_configuration(
+      bucket: bucket_name,
+      id: inventory_id,
+      inventory_configuration: inventory_configuration,
+      use_accelerate_endpoint: false
+    )
+  end
+
+  def inventory_configuration
+    filter_prefix = inventory_id
+    destination_prefix = File.join(INVENTORY_PREFIX, inventory_id)
+
+    if bucket_folder_path.present?
+      filter_prefix = File.join(bucket_folder_path, filter_prefix)
+      destination_prefix = File.join(bucket_folder_path, destination_prefix)
+    end
+
+    {
+      destination: {
+        s3_bucket_destination: {
+          bucket: "arn:aws:s3:::#{bucket_name}",
+          prefix: destination_prefix,
+          format: "CSV"
+        }
+      },
+      filter: {
+        prefix: filter_prefix
+      },
+      is_enabled: SiteSetting.enable_s3_inventory,
+      id: inventory_id,
+      included_object_versions: "Current",
+      optional_fields: ["ETag"],
+      schedule: {
+        frequency: "Daily"
```

Comment:
Because the inventory is only taken daily, our check for missing uploads should filter uploads that are created before the inventory file was created.